### PR TITLE
Abacus Transactions Protocol

### PR DESCRIPTION
**Pull request by jaredvu**
Utilize Abacus Transaction protocol to handle order placement and cancellation. Update loading states and analytics.

---

## Views

* `<MarketLinks>`
  * fix key, because `href` is possibly a duplicate key

* `<TradeBoxOrderView>`
  * enable conditional orders in production 

* `<HorizontalPanel>`
  * use `latestOrderStatus` to determine loading state

* `<OrderDetailsDialog>`/`<OrderActionsCell>`
  * fix handling of `cancelOrder` loading state

* `<TradeForm>`/`<ClosePositionForm>`
  *  use `latestOrderStatus` to determine loading state

* `<PlaceOrderButtonAndReceipt>`
  * button should be `disabled` while `isLoading` so users can't accidentally send multiple orders

## Constants/Types

* `constants/abacus`
  * add `AbacusTrackingProtocol` interface

* `constants/analytics`
  * add `TradeCancelOrder` event
  * add `TransferFaucetConfirmed` event

* `constants/localization`
  * delete error code string map

* `constants/trade`
  * deprecating use of `uncommitedOrderIds` so removing timeout

## Functions

* `lib/abacus/analytics`
  * add Analytics protocol to record events prompted by Abacus.

* `lib/abacus/dydxChainTransactions`
  * handle `placeOrder` and `cancelOrder` transactions using Abacus
  * moved logic from `useSubaccounts` to this Abacus protocol
  * add `setLocalWallet` method in order to create `subaccountClient`

* `lib/abacus`
  * use `analytics` protocol
  *  expose `placeOrder`, `cancelOrder`, and `closePosition` methods

* `lib/abacus/stateNotification`
  * save contents of `lastOrderChanged` method  to be used for loading states

## Hooks

* `hooks/useAccount`
  * set `localWallet` instead of `dydxAddress`

* `hooks/userSubaccount`
  * update `placeOrder` and `cancelOrder` to utilize Abacus transactions.
  * remove `placeOrderForSubaccount` and `cancelOrderForSubaccount` methods

## State
* `state/account`
  * add `latestOrder`

* `state/accountSelector
  * add selectors for `latestOrderId` and `latesterOrderStatus`
  
* `state/tradeMiddleware`/`state/_store`
  * deleted due to deprecation of `uncommitedOrderIds`

## Packages

* `@dydxprotocol/v4-abacus`
  * Consistent analytics events
  * Update the returned broadcastError stringKeys
  * updated: v0.4.20 -> v0.4.27

* `@dydxprotocol/v4-localization`
  * Update broadcastError stringKeys 
  * updated: v0.1.5 -> v0.1.7 

---

1. Testing order placement
  * Order placement should work
  * There should be a loading state on order placement with the loading state dismissing after the trade is confirmed.
  * Loading state on the horizontalPanel tabs should also display (except limit order GTT: orderStatus is never `pending`)

2. Testing order cancellation
  * Order cancel from the horizontalPanel should keep loading state and not flash (fixes ability to send multiple order cancels)
  * Order cancel from OrderDetailsDialog should also keep loading state until cancel is confirmed

### CODE REVIEW

File: src/lib/abacus/index.ts [hunk 1]
```diff
@@ -219,6 +228,17 @@ class AbacusStateManager {
...
+    this.stateManager.commitPlaceOrder(callback);
+
+  closePosition = (callback: (p0: boolean, p1: Nullable<ParsingError>) => void) =>
+    this.stateManager.commitClosePosition(callback);
+
+  cancelOrder = (orderId: string, callback: (p0: boolean, p1: Nullable<ParsingError>) => void) =>
+    this.stateManager.cancelOrder(orderId, callback);
+
   // ------ Utils ------ //
   placeOrderPayload = (): Nullable<HumanReadablePlaceOrderPayload> =>
```
Comment (rosepuppy) [inline src/lib/abacus/index.ts:243]: Are we able to deprecate these payload utils now?

File: src/lib/abacus/dydxChainTransactions.ts [hunk 2]
```diff
@@ -106,13 +123,120 @@ class DydxChainTransactions implements AbacusDYDXChainTransactionsProtocol {
...
+        reduceOnly,
+        triggerPrice ?? undefined
+      );
+
+      // Handle stateful orders
+      if ((tx as IndexedTx)?.code !== 0) {
+        throw new StatefulOrderError('Stateful order has failed to commit.', tx);
+      }
+
+      const hash = tx?.hash && Buffer.from(tx.hash).toString('hex').toUpperCase();
```
Comment (rosepuppy) [inline src/lib/abacus/dydxChainTransactions.ts:168]: hash is not used it, was it supposed to be logged?

File: src/lib/abacus/dydxChainTransactions.ts [hunk 3]
```diff
@@ -106,13 +123,120 @@ class DydxChainTransactions implements AbacusDYDXChainTransactionsProtocol {
...
+      return JSON.stringify(tx);
```
Comment (rosepuppy) [inline src/lib/abacus/dydxChainTransactions.ts:169]: probably ok here, should we use `parseToPrimitives` to be consistent here?

File: src/lib/abacus/dydxChainTransactions.ts [hunk 4]
```diff
@@ -106,13 +123,120 @@ class DydxChainTransactions implements AbacusDYDXChainTransactionsProtocol {
...
+      const tx = await this.compositeClient?.cancelOrder(
+        new SubaccountClient(this.localWallet, subaccountNumber),
+        clientId,
+        orderFlags,
+        clobPairId,
+        goodTilBlock ?? undefined,
+        goodTilBlockTime ?? undefined
+      );
+
+      return JSON.stringify(tx);
```
Comment (rosepuppy) [inline src/lib/abacus/dydxChainTransactions.ts:197]: probably ok here, should we use parseToPrimitives to be consistent here?

File: src/views/forms/TradeForm.tsx [hunk 5]
```diff
@@ -97,10 +99,11 @@ export const TradeForm = ({
   } = useTradeFormData();
 
   const { limitPrice, triggerPrice, trailingPercent } = price || {};
-  const hasUncommittedOrders = useSelector(calculateHasUncommittedOrders);
   const currentInput = useSelector(getCurrentInput);
   const { tickSizeDecimals, stepSizeDecimals } =
     useSelector(getCurrentMarketConfig, shallowEqual) || {};
+  const latestOrderStatus = useSelector(getLatestOrderStatus) ?? '';
+  const latestOrderId = useSelector(getLatestOrderId);
```
Comment (rosepuppy) [inline src/views/forms/TradeForm.tsx:106]: nit: is `latestOrderId` needed? It seems we only depend on `latestOrderStatus`

File: src/views/forms/ClosePositionForm.tsx [hunk 6]
```diff
@@ -82,8 +83,9 @@ export const ClosePositionForm = ({
   const { stepSizeDecimals } = useSelector(getCurrentMarketConfig, shallowEqual) || {};
   const { size: sizeData, summary } = useSelector(getInputClosePositionData, shallowEqual) || {};
   const { size, percent } = sizeData || {};
-  const hasUncommittedOrders = useSelector(calculateHasUncommittedOrders);
   const currentInput = useSelector(getCurrentInput);
+  const latestOrderStatus = useSelector(getLatestOrderStatus) ?? '';
+  const latestOrderId = useSelector(getLatestOrderId);
```
Comment (rosepuppy) [inline src/views/forms/ClosePositionForm.tsx:88]: nit: is `latestOrderId` needed? It seems we only depend on `latestOrderStatus`

**Review by rosepuppy — APPROVED** [review]
Generally LGTM! Mostly nits

File: src/views/forms/ClosePositionForm.tsx (hunk 6 above)
Comment (jaredvu) [reply src/views/forms/ClosePositionForm.tsx:88]: without `latestOrderId` as a value in the dep array, it is possible for the side effect to be skipped:
- the `latestOrderStatus` may not change between 2 different orders

File: src/lib/abacus/index.ts (hunk 1 above)
Comment (jaredvu) [reply src/lib/abacus/index.ts:243]: `placeOrderPayload` and `closePositionPayload` are still used

Comment (jaredvu) [reply src/lib/abacus/index.ts:243]: removed the `cancelOrderPayload` method

**Review by aforaleka — APPROVED** [review]
it seems like with this new implementation, there's an extra state between order pending and placed where we're showing an order diff state with outdated inputs -- is there any way we can avoid it? (e.g. clear the inputs?)

https://www.loom.com/share/5791e96f825349d1aa279c43e10450b5?sid=572b226d-b9f7-456c-9081-6f6fb9aa7017

File: src/lib/abacus/dydxChainTransactions.ts [hunk 7]
```diff
@@ -106,13 +130,138 @@ class DydxChainTransactions implements AbacusDYDXChainTransactionsProtocol {
...
+        postOnly,
+        reduceOnly,
+        triggerPrice,
+      } = params || {};
+
+      // Observe uncommitted order
+      this.store?.dispatch(addUncommittedOrderClientId(clientId));
+
+      setTimeout(() => {
+        this.store?.dispatch(removeUncommittedOrderClientId(clientId));
```
Comment (rosepuppy) [inline src/lib/abacus/dydxChainTransactions.ts:158]: I am a bit confused, it seems to me that `removeUncommittedOrderClientId` is only being called on timeout/error. Is uncommittedOrderClientId still being used?

Comment (rosepuppy) [reply src/lib/abacus/dydxChainTransactions.ts:158]: oh nvm, I see it being unset in `setLatestOrder`

**Review by rosepuppy — APPROVED** [review]
LGTM

File: src/state/account.ts [hunk 8]
```diff
@@ -86,6 +88,16 @@ export const accountSlice = createSlice({
     setTransfers: (state, action: PayloadAction<any>) => {
       state.transfers = action.payload;
     },
+    setLatestOrder: (state, action: PayloadAction<Nullable<SubaccountOrder>>) => {
```
Comment (jaredvu) [inline src/state/account.ts:91]: @rosepuppy When Abacus updates latestOrder we also remove the `clientId` from the `uncomittedOrderClientIds`.

**Pull request merged**